base/array.jl

```diff
@@ -696,12 +696,12 @@ _similar_for(c::AbstractArray, ::Type{T}, itr, ::HasShape, axs) where {T} =
     similar(c, T, axs)
 
 # make a collection appropriate for collecting `itr::Generator`
-_array_for(::Type{T}, ::SizeUnknown, ::Nothing) where {T} = Vector{T}(undef, 0)
-_array_for(::Type{T}, ::HasLength, len::Integer) where {T} = Vector{T}(undef, Int(len))
-_array_for(::Type{T}, ::HasShape{N}, axs) where {T,N} = similar(Array{T,N}, axs)
+_array_for2(::Type{T}, ::SizeUnknown, ::Nothing) where {T} = Vector{T}(undef, 0)
+_array_for2(::Type{T}, ::HasLength, len::Integer) where {T} = Vector{T}(undef, Int(len))
+_array_for2(::Type{T}, ::HasShape{N}, axs) where {T,N} = similar(Array{T,N}, axs)
 
 # used by syntax lowering for simple typed comprehensions
-_array_for(::Type{T}, itr, isz) where {T} = _array_for(T, isz, _similar_shape(itr, isz))
+_array_for(::Type{T}, itr, isz) where {T} = _array_for2(T, isz, _similar_shape(itr, isz))
```

Comment:
This is the main point of contention: JET can not rule out that this method ends up calling itself recursively, which then leads to follow-up weirdness (like it looking for a `length(::HasLength)` method, and of course failing to find one).

I am open for better names than `_array_for2` (perhaps `_array_for_inner` ?).

Reply:
Thanks for making this fix. I don't think `_array_for` and `_array_for2` really need to be generic functions with the same name, so I'm totally on board with separating them for type stability.
Maybe `_array_for_inner` would be better than `_array_for2`.

I'd like to add some tests using JET to prevent future regressions, but we lack the infra for that currently (and JET is currently not even compatible with nightly)